core/transform.h

```diff
@@ -57,108 +113,264 @@ constexpr auto operator*(const ValueAndVariance<T1> a, const T2 b) noexcept {
+    // this function for the iteration over each individual sparse_container.
+    // This then falls into case 2 and thus the recursion terminates with the
+    // second level.
+    if constexpr (is_sparse_v<decltype(vals[0])>) {
+      op(ValuesAndVariances{vals[i], vars[i]},
+         value_and_maybe_variance(other, i)...);
+    } else {
+      ValueAndVariance _{vals[i], vars[i]};
+      op(_, value_and_maybe_variance(other, i)...);
+      vals[i] = _.value;
+      vars[i] = _.variance;
+    }
+  }
+}
 
-template <class T> struct as_view {
-  using value_type = typename T::value_type;
-  bool hasVariances() const { return data.hasVariances(); }
-  auto values() const { return data.valuesView(dims); }
-  auto variances() const { return data.variancesView(dims); }
+template <class Op, class T, class... Ts>
+void transform_in_place_impl(Op op, T &&vals, Ts &&... other) {
+  for (scipp::index i = 0; i < scipp::size(vals); ++i)
+    op(vals[i], other[i]...);
+}
 
-  T &data;
-  const Dimensions &dims;
+/// Broadcast a constant to arbitrary size. Helper for TransformSparse.
+///
+/// This helper allows the use of a common transform implementation when mixing
+/// sparse and non-sparse data.
+template <class T> struct broadcast {
+  constexpr auto operator[](const scipp::index) const noexcept { return value; }
+  T value;
 };
-template <class T> as_view(T &data, const Dimensions &dims)->as_view<T>;
+template <class T> broadcast(T)->broadcast<T>;
 
+/// Functor for implementing operations with sparse data.
+///
+/// This is (conditionally) added to an overloaded set of operators provided by
+/// the user. If the data is sparse the overloads by this functor will match in
+/// place of the user-provided ones. We then recursively call the transform
+/// function. In this second call we have descended into the sparse container so
+/// now the user-provided overload will match directly.
+template <class Op> struct TransformSparse {
+  Op op;
+  template <class T> constexpr void operator()(sparse_container<T> &x) const {
+    transform_in_place_impl(op, x);
+  }
+  template <class T> constexpr void operator()(ValuesAndVariances<T> x) const {
+    transform_in_place_with_variance_impl(op, x);
+  }
+  template <class T1, class T2>
+  constexpr void operator()(sparse_container<T1> &a, const T2 b) const {
+    transform_in_place_impl(op, a, broadcast{b});
+  }
+  template <class T1, class T2>
+  constexpr void operator()(ValuesAndVariances<T1> a, const T2 b) const {
+    transform_in_place_with_variance_impl(op, a, broadcast{b});
+  }
+  template <class T1, class T2>
+  constexpr void operator()(sparse_container<T1> &a,
+                            const sparse_container<T2> &b) const {
+    expect::sizeMatches(a, b);
+    transform_in_place_impl(op, a, b);
+  }
+  template <class T1, class T2>
+  constexpr void operator()(ValuesAndVariances<T1> a,
+                            const sparse_container<T2> b) const {
```

Comment:
```suggestion
                            const sparse_container<T2> &b) const {
```

Reply:
:+1, but this overload is removed in a follow-up branch, not changing here.